toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,91 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(bool first_name_component,
+                                              llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  auto& sem_ir = file_context_.sem_ir();
+  auto names = sem_ir.names();
+  auto types = sem_ir.types();
+  while (name_scope_id.is_valid()) {
+    const auto& parent = sem_ir.name_scopes().Get(name_scope_id);
+    if (parent.inst_id == SemIR::InstId::PackageNamespace) {
+      break;
+    }
+    if (!first_name_component) {
+      os << '.';
+    }
+    first_name_component = false;
+    CARBON_KIND_SWITCH(sem_ir.insts().Get(parent.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        const auto& impl = sem_ir.impls().Get(impl_decl.impl_id);
+        if (auto opt_class_self =
+                types.TryGetAs<SemIR::ClassType>(impl.self_id)) {
+          MangleInverseQualifiedNameScope(
+              true, os,
+              sem_ir.classes().Get(opt_class_self->class_id).scope_id);
+        } else {
+          auto builtin_self = types.GetAs<SemIR::BuiltinInst>(impl.self_id);
+          os << builtin_self.builtin_inst_kind.label();
+        }
+        os << ':';
+        auto opt_interface_constraint =
+            types.GetAs<SemIR::InterfaceType>(impl.constraint_id);
+        os << names.GetFormatted(sem_ir.interfaces()
+                                     .Get(opt_interface_constraint.interface_id)
+                                     .name_id);
+        return;
+      }
+      case CARBON_KIND(SemIR::ClassDecl class_decl): {
+        auto& class_node = sem_ir.classes().Get(class_decl.class_id);
+        os << names.GetFormatted(class_node.name_id);
+        break;
+      }
+      case SemIR::Namespace::Kind: {
+        auto name = names.GetAsStringIfIdentifier(parent.name_id);
+        CARBON_CHECK(name) << "Unexpected special name for namespace: "
+                           << parent.name_id;
```

Comment:
After our whole discussion, I was thinking you'd do `GetAsStringIfIdentifier` everywhere, were you running into an issue or did I misunderstand?

Reply:
Yeah, think I just got a bit confused after stumbling over the NameScope::name_id is only valid for namespaces, and other name_ids you have to go looking in the underlying entity - and then I was looking for how to render those entities' names and came across other APIs and copied those, unsure if/how they were load bearing.

But thanks for the prompt - zooming out again, with the code tidied up a bit, I've been able to use that consistently everywhere, I think. (though not able to share many call sites/deduplicate much code, since the name_ids all come from different places, etc)